gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/testing/LocalResourceManagerHelper.java

```diff
@@ -0,0 +1,588 @@
+          case "pageToken":
+            // support pageToken when Cloud Resource Manager supports this (#421)
+            break;
+          case "pageSize":
+            // support pageSize when Cloud Resource Manager supports this (#421)
+            break;
+        }
+      }
+    }
+    return options;
+  }
+
+  private static final boolean isValidProject(Project project) {
+    if (project.getProjectId() == null) {
+      log.info("Project ID cannot be empty.");
+      return false;
+    }
+    if (!isValidIdOrLabel(project.getProjectId(), 6, 30)) {
+      log.info("Project " + project.getProjectId() + " has an invalid ID."
+          + " See https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects"
+          + " for more information.");
+      return false;
+    }
+    if (project.getName() != null) {
+      for (char c : project.getName().toCharArray()) {
+        if (!PERMISSIBLE_PROJECT_NAME_PUNCTUATION.contains(c) && !Character.isLetterOrDigit(c)) {
+          log.info("Project " + project.getProjectId() + " has an invalid name."
+              + " See https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects"
+              + " for more information.");
+          return false;
+        }
+      }
+    }
+    if (project.getLabels() != null) {
+      if (project.getLabels().size() > 256) {
+        log.info("Project " + project.getProjectId() + " exceeds the limit of 256 labels.");
+        return false;
+      }
+      for (Map.Entry<String, String> entry : project.getLabels().entrySet()) {
+        if (!isValidIdOrLabel(entry.getKey(), 1, 63)
+            || !isValidIdOrLabel(entry.getValue(), 0, 63)) {
+          log.info("Project " + project.getProjectId() + " has an invalid label entry."
+              + " See https://cloud.google.com/resource-manager/reference/rest/v1beta1/projects"
+              + " for more information.");
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private static final boolean isValidIdOrLabel(String value, int minLength, int maxLength) {
+    for (char c : value.toCharArray()) {
+      if (c != '-' && !Character.isDigit(c)
+          && (!Character.isLetter(c) || !Character.isLowerCase(c))) {
+        return false;
+      }
+    }
+    if (value.length() > 0 && (!Character.isLetter(value.charAt(0)) || value.endsWith("-"))) {
+      return false;
+    }
+    return value.length() >= minLength && value.length() <= maxLength;
+  }
+
+  Response create(Project project) throws IOException {
+    project.setLifecycleState("ACTIVE");
+    project.setProjectNumber(Math.abs(PROJECT_NUMBER_GENERATOR.nextLong()));
+    project.setCreateTime(ISODateTimeFormat.dateTime().print(new Date().getTime()));
+    Response response;
+    if (!isValidProject(project)) {
+      response = Error.INVALID_ARGUMENT.response;
+    } else if (projects.containsKey(project.getProjectId())) {
+      response = Error.ALREADY_EXISTS.response;
+      log.info(
```

Comment:
Same comment, we should prefer communicating request level errors via the error response rather than using the logs.